src/core/execute-serialize.c

```diff
@@ -1117,20 +1117,40 @@ static int exec_runtime_serialize(const ExecRuntime *rt, FILE *f, FDSet *fds) {
                         return r;
 
                 if (rt->shared->netns_storage_socket[0] >= 0 && rt->shared->netns_storage_socket[1] >= 0) {
-                        r = serialize_fd_many(f, fds, "exec-runtime-netns-storage-socket", rt->shared->netns_storage_socket, 2);
+                        r = serialize_fd_many_full(
+                                        f,
+                                        fds,
+                                        index ? "exec-runtime-netns-storage-socket-by-fd-index" :
+                                                "exec-runtime-netns-storage-socket",
+                                        rt->shared->netns_storage_socket,
+                                        2,
+                                        index);
                         if (r < 0)
                                 return r;
                 }
 
                 if (rt->shared->ipcns_storage_socket[0] >= 0 && rt->shared->ipcns_storage_socket[1] >= 0) {
-                        r = serialize_fd_many(f, fds, "exec-runtime-ipcns-storage-socket", rt->shared->ipcns_storage_socket, 2);
+                        r = serialize_fd_many_full(
+                                        f,
+                                        fds,
+                                        index ? "exec-runtime-ipcns-storage-socket-by-fd-index" :
+                                                "exec-runtime-ipcns-storage-socket",
+                                        rt->shared->ipcns_storage_socket,
+                                        2,
+                                        index);
```

Comment:
please rework this so that fd vs. index is hidden inside of serialize_fd() and serialize_fd_many(), i.e. that we either serialize a literal number for fd, and for index nr serialize `#` followed by the index. that makes things a lot simpler, as the difference is mostly abstracted away, and you don't need different field names anymore.

Reply:
As already mentioned, I think making the parsing type dependent on the value instead of the key is a mistake. It makes everything more complicated, not less, because suddenly (and unlike every other field) the key is no longer enough to know what to do with a value, but parsing through the value is needed to know how to further parse it, which for example adds corner cases like mixing both in the same message, which would be nonsensical but that things like the fuzzer would no doubt start firing at us so it would have to be handled, making everything even more complicated. So I am going to leave this as it is.

Reply:
i remain unconvinced. if you want to convince me show me an excerpt on how it would look like where it would get worse.

and even if, there's no excuse for the ternary op of the field name in each and every invocation of the fd serializer, because you could just suffix it internally...

also if you bind the logic to discern both cases to the key then the fuzzer would still throw serialization with both at us, which we need to handle too. there's not much difference there.

Reply:
> and even if, there's no excuse for the ternary op of the field name in each and every invocation of the fd serializer, because you could just suffix it internally...

Sure, moved the suffixing to the various helpers

> also if you bind the logic to discern both cases to the key then the fuzzer would still throw serialization with both at us, which we need to handle too. there's not much difference there.

Not really, it's one or the other right now, it doesn't even require any special handling. If you get an index but only have a set, deserializing will try to use an empty array and return einval straight away - as it should in any case, and viceversa.

Reply:
i don't get it, why does deserializing ever operate on an array? fdset should be just fine for that too